README.md
[![Build Status](https://travis-ci.org/kaleguy/simpleweatherproxy.svg?branch=master)](https://travis-ci.org/kaleguy/simpleweatherproxy)
[![Dependency Status](https://david-dm.org/kaleguy/simpleweatherproxy.svg)](https://david-dm.org/kaleguy/simpleweatherproxy)
[![Coverage Status](https://coveralls.io/repos/github/kaleguy/simpleweatherproxy/badge.svg?branch=master)](https://coveralls.io/github/kaleguy/simpleweatherproxy?branch=master)

## Synopsis

A proxy to provide a basic API to scrape JSON data from ResearchGate pages.  

Endpoints implemented:

* Article info by title
* Author info
* List of articles by author
* List of citations by article

Example url:
```
http://localhost:8888/rg/article?title=282790843_Differences_in_Media_Preference_Mediate_the_Link_Between_Personality_and_Political_Orientation
```

The existing endpoints can be easily adapted to scrape a different site.

Implemented with Restify (version 7).

## Update for Version 1.1

ResearchGate has made changes to their site that make scraping more of a challenge, so version 1.1 has a few changes:

1. Author lists on the article page may be incomplete and require another cick to retrieve the full list. 
  No fix for this currently. If you scrape a page like this, you'll need to add the additional authors manually.
1. Some articles have different formatting than others. To deal with this, you can now add selectors (see below) in an 
  array, the selector that returns a value will be used.
1. Some items, for example number of reads no longer have any enclosing element and can't be selected by CSS alone. 
  For this case, the feature has been added of being able to append a regular expression to the end of the selector.
  See below for example.
1. In some cases ResearchGate may not return a result unless you send a cookie. If this happens, access the site via 
  browser and then retrieve the cookie from the Developer Console of your browser, and paste that value into the 
  .set("Cookie", '') line in scraper.js      


## Configuring Endpoints

To scrape the page referenced above, the following selectors are configured in the endpoint:

```
    const selectors = {
      pubdate: 'meta[property="citation_publication_date"]',
      title: 'h1.nova-e-text--size-xxxl',
      citations:
        [
          'DIV.nova-e-text.nova-e-text--size-m.nova-e-text--family-sans-serif.nova-e-text--spacing-none.nova-e-text--color-inherit.publication-resources-summary__see-all-count strong',
          '.ga-resources-citations span.publication-resource-link-amount'
        ],
      references: [
        'DIV.nova-o-pack__item:nth-child(2) DIV.nova-e-text.nova-e-text--size-m.nova-e-text--family-sans-serif.nova-e-text--spacing-none.nova-e-text--color-inherit.publication-resources-summary__see-all-count strong',
        '.ga-resources-references span.publication-resource-link-amount'
        ],
      reads: [
        'DIV.nova-e-text.nova-e-text--size-m.nova-e-text--family-sans-serif.nova-e-text--spacing-xxs.nova-e-text--color-grey-700:not(span) /with\\s(.*?)$/',
        '.publication-meta-stats',
      ],
      journal:
        [
        'A.nova-e-link.nova-e-link--color-blue.nova-e-link--theme-bare',
        '.publication-meta-journal A',
        ],
      abstract: [
         '.publication-abstract .nova-e-text--spacing-auto',
         'DIV.nova-e-text.nova-e-text--size-m.nova-e-text--family-sans-serif.nova-e-text--spacing-auto.nova-e-text--color-inherit'
        ],
      authors: {
        selector: '.publication-author-list__item',
        subselectors: [
          { name: '.nova-v-person-list-item__title A' },
          { institution: 'LI.nova-v-person-list-item__meta-item:last-child SPAN' }
        ]
      }
    }

```

Note that some selectors (e.g. 'abstract') are arrays. This is to deal with multiple page formats.

Also note the regular expression at the end of the first "reads" selector. The value of this applied to the value of the 
first selector will be returned.


Sample output for the above url:

```
{
   "pubdate":"2015/10/01",
   "title":"Differences in Media Preference Mediate the Link Between Personality and Political Orientation",
   "cits":"6",
   "refs":"62",
   "date":"2015-10",
   "reads":"2,632 Reads",
   "journal":{
      "href":"journal/1467-9221_Political_Psychology",
      "text":"Political Psychology"
   },
   "abstract":"Research has consistently demonstrated that political liberalism is predicted by the personality trait Openness to Experience and conservatism by trait Conscientiousness. Less well studied, however, is how trait personality influences political orientation. The present study investigated whether differences in media preference might mediate the links between personality and political orientation. Participants completed measures of Big Five personality, media preferences, and political orientation. Results revealed that increased preferences for Dark/Alternative and Aesthetic/Musical media genres, as well as decreased preferences for Communal/Popular media genres, mediated the association between Openness to Experience and liberalism. In contrast, greater preferences for Communal/Popular and Thrilling/Action genres, as well as lower preferences for Dark/Alternative and Aesthetic/Musical genres mediated the link between Conscientiousness and conservatism.",
   "authors":[
      {
         "name":{
            "href":"https://www.researchgate.net/profile/Xiaowen_Xu4",
            "text":"Xiaowen Xu"
         },
         "institution":"University of Toronto"
      },
      {
         "name":{
            "href":"https://www.researchgate.net/profile/Jordan_Peterson2",
            "text":"Jordan B Peterson"
         },
         "institution":"University of Toronto"
      }
   ],
   "url":"https://www.researchgate.net/publication/282790843_Differences_in_Media_Preference_Mediate_the_Link_Between_Personality_and_Political_Orientation"
}
```
 
## More About Selectors

The following examples show how the selectors work with the above target url.

The selector object takes keys with CSS selector values. For example:

```
pubdate: 'meta[property="citation_publication_date"]'
```

This will result in JSON with the same key, with the value being the text found at the target document at the specified location.

If the CSS selector returns an 'A' element, the value returned will be an object, e.g. :

```
journal: '.publication-meta-journal A',
```

returns:

```
  "journal":{
      "href":"journal/1467-9221_Political_Psychology",
      "text":"Political Psychology"
   },
```

If the selector value is an object, then a list of objects will be returned. For example the
following will return an array of article objects (the publicaton-author-list__item selector selects for an article). 
The subselectors determine the keys of the list object. 

```
  authors: {
    selector: '.publication-author-list__item',
    subselectors: [
      { name: '.nova-v-person-list-item__title A' },
      { rating: '.nova-v-person-list-item__meta SPAN:first-child' },
      { institution: '.nova-v-person-list-item__meta LI:nth-child(2) SPAN' }
    ]
  }  
```

returns:

```
   "authors":[
      {
         "name":{
            "href":"https://www.researchgate.net/profile/Xiaowen_Xu4",
            "text":"Xiaowen Xu"
         },
         "rating":"14.72",
         "institution":"University of Toronto"
      },
      {
         "name":{
            "href":"https://www.researchgate.net/profile/Jordan_Peterson2",
            "text":"Jordan B Peterson"
         },
         "rating":"39.79",
         "institution":"University of Toronto"
      }
   ],
```    

## XPath Selectors

If a selector begins with '//' it will be assumed to be an XPath selector and not a CSS selector. Element names need to be namespaced with "x", for example:

```
'//x:th[contains(text(),\'Document type\')]/parent::*/x:td'
```

## Multipage Targets

Endpoints can be configured to scrape multiple pages at a given target site.

Endpoints use the 'scaper' function with a path for the target. If the path is an array, the scraper will visit each path and return the merged results.

See the rg.js file for examples.


## Installation

Note: if you want the weather demo endpoint to be functional,
rename .config.example.json to .config.json, edit with your openWeather API key.

To install and run, install and start mongodb. Then:

```
npm install
npm start
```

## Caching

If mongodb is running locally, the results of searches will get cached to Mongo.

## Swagger-JSDoc and Swagger-Test

This project uses the swagger-jsdoc module, which allows for documenting endpoints inline with Swagger YAML:
(The example below below is for the example Open Weather API endpoint.

```
  /**
   * @swagger
   * /weather/{city}:
   *   get:
   *     description: "Returns current weather in the specified city to the caller"
   *     summary: "Get Weather by City"
   *     tags: [Weather]
   *     parameters:
   *       - name: city
   *         in: path
   *         description: "The city you want weather for in the form city,state,country"
   *         required: true
   *         type: string
   *     produces:
   *       - application/json
   *     responses:
   *       200:
   *         description: |-
   *           weather response
   *         schema:
   *           $ref: "#/definitions/report"
   *     x-amples:
   *     - description: Get Weather
   *       request:
   *         params:
   *          city: NYC
   *       responses:
   *          200:
   *            headers:
   *               content-type: "application/json; charset=utf-8"
   *            body:
   *               name: 'New York'
   *
   */
```
The swagger-jsdoc module converts this into swagger json, which you can then view with the Swagger UI webapp, see next
section.

The x-amples section in the comments above will be turned into a test by swagger-test, see the testing section below.

## API Reference

Swagger Docs generated from inline Swagger comments, processed by Swagger-JSDocs. 

When running locally the Swagger UI will be reachable at http://localhost:8888/public/swagger/index.html


## Tests

Tests generated from Swagger Docs (also runs linter):
```
npm run test-api
```

There are also a couple more test suites for reference.
Jasmine tests:

```
npm test
```

Tape/Supertest tests:
```
npm run test-agent
```

## Run with Docker

```
docker build -t rg .
docker run -p 8888:8888 rg  
```

## Coverage

Create coverage report with Istanbul, send to Coveralls to keep a record.
```
npm run coverage
```

## License

MIT


Changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
* [Swagger UI](http://scrape-rg-dev.us-east-1.elasticbeanstalk.com/public/swagger/index.html)


## [1.2.0] - 2019-01-17
### Added
- XPath selectors

## [1.1.0] - 2018-12-21
### Added
- Selectors with Regex
- Set Cookie
- Selector Arrays
- Mongodb cache


tests/weather-spec.js
describe('OpenWeatherMap API', function(){

  const preq = require('preq');
  const server = require('../index.js');
  const serverOptions = {
    uri: 'http://localhost:8888/weather/',
    headers: {
      'Accept': 'application/json'
    }
  };


  it('Get the correct city name returned', function (done){

    serverOptions.uri += 'new york';
    preq.get(serverOptions)
      .then(function(res){
         console.log('Response: ', serverOptions.uri, res.body);
         expect(res.body.name).toBe('New York');
       //  server.close();
         done();
      });

  });

  it('Get the correct city name returned on text request', function (done){
    serverOptions.headers.Accept = 'text/plain';
    preq.get(serverOptions)
      .then(function(res){
        console.log('Response: ', serverOptions.uri, res);
        const body = JSON.parse(res.body);
        expect(body.name).toBe('New York');
        server.close();
        done();
      });

  });

/*  afterAll(function() {
    console.log('afterAll');
    server.close();
  });*/

});


apitests/swagger-test.js
'use strict';

// mocha defines to avoid JSHint breakage
/* global describe, it, before, beforeEach, after, afterEach */

var swaggerTest = require('swagger-test');
var preq = require('preq');

var chai = require('chai');
chai.use(require('chai-shallow-deep-equal'));

var assert = chai.assert;

var xamples;

describe('specification-driven tests setup', function () {

  before(function(done){
    preq.get({
        uri: 'http://localhost:8888/api-docs.json',
        headers: {
          'Accept': 'application/json'
        }
      })
      .then(function(res){
        xamples = res.body;
        xamples = swaggerTest.parse(xamples);
        tests();
        done();
      })
  });

  // need this to allow tests() to run
  it('dummy', function(){
    return true;
  });


});

function tests(){
  describe('specification-driven tests', function() {
    xamples.forEach(function (xample) {
      it(xample.description, function () {
        xample.request.headers =   {'Accept': 'application/json'};
        return preq[xample.request.method](xample.request)
          .then(function (response) {
              assert.shallowDeepEqual(response, xample.responses[response.status]);
          });
      });
    });
  });
}


api-tests/swagger-test.js
'use strict';

// mocha defines to avoid JSHint breakage
/* global describe, it, before, beforeEach, after, afterEach */

const server = require('../index.js');
const swaggerTest = require('swagger-test');
const preq = require('preq');

const chai = require('chai');
chai.use(require('chai-shallow-deep-equal'));

const assert = chai.assert;

let xamples;

describe('specification-driven tests setup', function () {

  before(function(done){
    preq.get({
        uri: 'http://localhost:8888/api-docs.json',
        headers: {
          'Accept': 'application/json'
        }
      })
      .then(function(res){
        xamples = res.body;
        xamples = swaggerTest.parse(xamples);
        tests();
        done();
      })
  });

  // need this to allow tests() to run
  it('dummy', function(){
    return true;
  });


});

function tests(){
  describe('specification-driven tests', function() {
    xamples.forEach(function (xample) {
      it(xample.description, function () {
        if (! xample.request.headers) {
          xample.request.headers = {'Accept': 'application/json'};
        }
        return preq[xample.request.method](xample.request)
          .then(function (response) {
            xample.finished = true;
            assert.shallowDeepEqual(response, xample.responses[response.status]);
            checkXamplesFinished();
          });
      });
    });
  });
}

function checkXamplesFinished(){

  let finished = true;
  xamples.forEach(function(xample){
    if (! xample.finished){
      finished = false;
    }
  });
  if (finished){
    server.close();
  }

}


index.js
/*jslint node: true, indent: 2 */
'use strict'

const corsMiddleware = require('restify-cors-middleware')
const swaggerJSDoc = require('swagger-jsdoc')
var exports = module.exports = {}

const MongoClient = require('mongodb').MongoClient;
const assert = require('assert');

const url = process.env.MONGO_URL || 'mongodb://localhost:27017';
let db = null;
let cache = null;

const options = {
  swaggerDefinition: {
    info: {
      title: 'Scraper API for ResearchGate', // Title (required)
      description: 'Endpoints for extracting JSON data from ResearchGate endpoints',
      version: '1.0.0' // Version (required)
    },
    tags: [
      {
        "name": "Weather",
        "description": "Simple pass-through demo endpoint."
      },
      {
        "name": "ResearchGate",
        "description": "Endpoints to extract JSON from ResearchGate pages."
      }
    ],
    basePath: '',
    schemes: ['http', 'https']
  },
  apis: [
    __dirname + '/routes/weather.js',
    __dirname + '/routes/rg.js',
    __dirname + '/index.js'
  ] // Path to the API docs
}
const swaggerSpec = swaggerJSDoc(options)

/**
 *
 * @swagger
 * definitions:
 *   report:
 *     type: object
 *     required:
 *       - coord
 *       - weather
 *       - base
 *       - main
 *       - visibility
 *       - wind
 *       - clouds
 *       - dt
 *       - sys
 *       - id
 *       - name
 *       - cod
 *     properties:
 *       coord:
 *         type: object
 *       weather:
 *         type: array
 *         items:
 *           type: string
 *       base:
 *         type: string
 *       main:
 *         type: object
 *       visibility:
 *         type: integer
 *       wind:
 *         type: object
 *       clouds:
 *         type: object
 *       dt: integer
 *       sys:
 *         type: object
 *       id: integer
 *       name: string
 *       cod: integer
 *
 */


let restify, bunyan, routes, log, server

restify = require('restify')
bunyan  = require('bunyan')
routes  = require('./routes/')

log = bunyan.createLogger({
  name        : 'restify',
  level       : process.env.LOG_LEVEL || 'info',
  stream      : process.stdout,
  serializers : bunyan.stdSerializers
})

server = restify.createServer({
  name : 'restify',
  log  : log,
})

MongoClient.connect(url, function(err, client) {

  // assert.equal(null, err);
  if (err) {
    console.log('No database available, results will not be cached.')
    db = {}
    cache = {}
    cache.findOne = cache.insertOne = (o, cb) => { return cb() }
  } else {
    console.log('Connected successfully to server')
    db = client.db('rg-cache')
    cache = db.collection('documents');
  }

  server.use(corsMiddleware({
    origins: [
      'https://kaleguy.github.io',
      'http://localhost:8080'
    ]// defaults to ['*']
    //  credentials: true,                 // defaults to false
    // headers: ['x-foo']                 // sets expose-headers
  }).actual)
  server.use(restify.plugins.bodyParser({ mapParams: false }))
  server.use(restify.plugins.queryParser())
  server.use(restify.plugins.gzipResponse())
  server.pre(restify.plugins.pre.sanitizePath())

  // This version of restify can't handle hyphens in url params... gross hack here to fix
  function encodeHyphens(req, res, next) {
    req.url = req.url.replace(/-/g, '__');
    return next();
  }
  // server.pre(encodeHyphens)

  /*jslint unparam:true*/
  // Default error handler. Personalize according to your needs.
  /* istanbul ignore next */
  server.on('uncaughtException', function (req, res, route, err) {
    console.log('******* Begin Error *******')
    console.log(route)
    console.log('*******')
    console.log(err.stack)
    console.log('******* End Error *******')
    if (!res.headersSent) {
      return res.send(500, { ok : false })
    }
    res.write("\n")
    res.end()
  })
  /*jslint unparam:false*/

  // server.on('after', restify.plugins.auditLogger({ log: log }))
  routes(server)

  // serve swagger docs
  server.get('/public/swagger/*', restify.plugins.serveStatic({
    directory: __dirname,
    default: 'index.html'
  }))

  // serve swagger spec
  server.get('/api-docs.json', function(req, res) {
    res.setHeader('Content-Type', 'application/json')
    res.send(swaggerSpec)
  })

  console.log('Server started.')
  server.listen(process.env.PORT || 8888, function () {
    log.info('%s listening at %s', server.name, server.url)
  })
  const app = server

  function checkCache(req, res, next) {
    const cb = (err, result) => {
      if (result) {
        delete result._id;
        console.log(`Retrieving ${result.key} from cache.`)
        res.json(result);
      } else {
        console.log(`Retrieving ${req.url}.`)
        next();
      }
    }
    cache.findOne( { key : req.url }, cb );
    //next()
  }
  app.use(checkCache);

  module.exports = app
  exports.cache = cache
  exports.close = server.close

});





lib/scraper.js
const assert = require('assert')
const request = require('superagent')
const { JSDOM } = require('jsdom')
const _ = require('lodash')
const moment = require('moment')
const app = require('../index')
const xpath = require('xpath');
const parse5 = require('parse5');
const xmlser = require('xmlserializer');
const xmlDom = require('xmldom').DOMParser;

function mergeList (d) {
  const merged = {}
  d.forEach(collection => {
    _.each(collection, (v, k) => {
      if (_.isArray(merged[k]) && _.isArray(v)) {
        merged[k] = _.concat(merged[k], v)
      } else {
        merged[k] = v
      }
    })
  })
  return merged
}

module.exports = function (url) {
  let module = {};

  function cache(d) {
    return new Promise((resolve, reject) => {
      cb = () => {
        return resolve(d)
      }
      d.cached = new Date();
      app.cache.insertOne(d, cb);
    });
  }

  /**
   *
   * @param selectors {object} Object with css selectors for plucking data
   * @param path {string or array} e.g. '/publication/{id}/JordanPeterson' or ['/publication/{id}/JordanPeterson','/publication/{id}/JordanPeterson2'], the subpath in the target API for a particular endpoint
   * @param res {object} the restify response object
   */
  module.scrape = function scrape (res, selectors, path, id, key) {

    if (_.isArray(path)) {
      let promises = [];
      path.forEach(p => {
        promises.push(getResultSet(p, selectors, id))
      })
      Promise.all(promises)
        .then(d => {
          d = mergeList(d)
          d.key = key
          return cache(d)
        })
        .then(d => {
        return res.json(d)
      }).catch(e => res.send(e.message))
    } else {
      getResultSet(path, selectors, id)
        .then(d => {
          d.key = key;
          return cache(d);
        })
        .then(d => res.json(d))
        .catch(e => res.json(e.message))
    }

  }

  function getResultSet (path, selectors, id) {
    path = path.replace('{id}', id)
    path = path.replace(/%27/g, '\'')
    const uaString = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36'
    return new Promise((resolve, reject) => {
      request
        .get(url + path)
        .set('accept', 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8')
        .set('user-agent', uaString)
        .set('accept-encoding', 'gzip, deflate, br')
        .set('Cookie', process.env.COOKIE || '')
        .set('accept-language', 'en-US,en;q=0.9')
        .end((err, xres) => {
          if (err) {
            console.log('ERROR', url + path, err.message);
            resolve([])
          }
          console.log('Retrieved data, processing...')
          const dom = new JSDOM(xres.text)
          const data = {}
          getDataFromSelectors(dom, selectors, data, xres.text)
          console.log('Returning processed data.')
          data.url = url + path
          resolve(data)
          // Calling the end function will send the request
        });
    })
  }

  function getDomList (dom, selector) {
    const items = []
    let els = dom.querySelectorAll(selector.selector)
    const subselectors = selector.subselectors
    els.forEach(el => {
      // console.log(el)
      let item = {}
      _.each(subselectors, (s) => {
        let t = ''
        let propName = Object.keys(s)[0]
        let propSelector = s[propName]
        if (_.isObject(propSelector)) {
          item[propName] = getDomList(el, propSelector)
          return
        }
        if (el.querySelector(propSelector)) {
          t = getDomText(el, propSelector)
        }
        if (t) {
          item[propName] = t
        }
      })
      if (!_.isEmpty(item)) {
        items.push(item)
      }
    })
    return items
  }

  function getDomText (dom, selector, docText) {
    let text = ''
    if (_.isObject(selector)) {
      return getDomList(dom, selector)
    }
    // if the selector is xpath
    let el = {
      textContent: ''
    }
    if (/^\/\//.test(selector)) {
       const document = parse5.parse(docText);
       const xhtml = xmlser.serializeToString(document);
       const doc = new xmlDom().parseFromString(xhtml);
       const select = xpath.useNamespaces({'x': 'http://www.w3.org/1999/xhtml'});
       const nodes = select(selector, doc);
       if (nodes.length) {
         el = nodes[0]
       }
    } else {
      el = null
    }

    // remove regex from selector
    const slashIndex = selector.indexOf('/')
    let regex = ''
    if (slashIndex > 0) {
      regex = selector.substr(slashIndex)
      regex = regex.substr(1)
      regex = regex.substr(0, regex.length - 1)
      regex = new RegExp(regex)
      selector = selector.substr(0, slashIndex)
      console.log(selector, regex)
    }
    el = el || dom.querySelector(selector)
    if (el) {
      text = el.textContent.replace(/\s+·\s+/,'').trim() // hack for garbage in rg text
      if (/^meta/.test(selector)) {
        text = el.getAttribute('content').trim()
      }
      if (regex) {
         const match = text.match(regex)
         if (match && match[1]) {
           text = match[1]
         }
      }
      if (el.tagName === 'A') {
        text = {
          href: el.getAttribute('href'),
          text: text
        }
      }
    }
    return text
  }
  function getDataFromSelectors (dom, selectors, data, text) {
    _.each(selectors, (v, k) => {
      if (_.isArray(v)) {
         _.each(v, (vv) => {
            const o = {}
            o[k] = vv
            getDataFromSelectors(dom, o, data)
         })
      }
      let t = getDomText(dom.window.document, v, text)
      if ((k === 'date') && t) { // TODO: HACK
        try {
          t = moment(t).format('YYYY-MM')
        } catch (e) {

        }
      }
      if (!data[k]) {
        data[k] = t
      }
    })
  }
  return module
}


routes/rg.js

const nconf = require('nconf')

nconf.argv()
  .env()
  .file({ file: '.config.json' })

module.exports = function (server) {

  const scraper = require('../lib/scraper')('https://www.researchgate.net')

  /**
   * @swagger
   * /rg/article:
   *   get:
   *     description: "Returns ResearchGate article info."
   *     summary: "Get ResearchGate article info"
   *     tags: [ResearchGate]
   *     parameters:
   *       - name: title
   *         in: query
   *         description: "The ResearchGate article title"
   *         required: true
   *         default: 304662727_Ideological_Reactivity_Political_Conservatism_and_Brain_Responsivity_to_Emotional_and_Neutral_Stimuli
   *         type: string
   *         consumes:
   *           - application/json
   *           - application/text
   *     produces:
   *       - application/json
   *     responses:
   *       200:
   *         headers:
   *           content-type: "text/plain"
   */
  server.get('/rg/article', function (req, res, next) {
    const id = req.query.title
    const selectors = {
      pubdate: 'meta[property="citation_publication_date"]',
      title: 'h1.nova-e-text--size-xxxl',
      citations:
        [
          'DIV.nova-e-text.nova-e-text--size-m.nova-e-text--family-sans-serif.nova-e-text--spacing-none.nova-e-text--color-inherit.publication-resources-summary__see-all-count strong',
          '.ga-resources-citations span.publication-resource-link-amount'
        ],
      references: [
        'DIV.nova-o-pack__item:nth-child(2) DIV.nova-e-text.nova-e-text--size-m.nova-e-text--family-sans-serif.nova-e-text--spacing-none.nova-e-text--color-inherit.publication-resources-summary__see-all-count strong',
        '.ga-resources-references span.publication-resource-link-amount'
        ],
      reads: [
        'DIV.nova-e-text.nova-e-text--size-m.nova-e-text--family-sans-serif.nova-e-text--spacing-xxs.nova-e-text--color-grey-700:not(span) /with\\s(.*?)$/',
        '.publication-meta-stats',
      ],
      journal:
        [
        'A.nova-e-link.nova-e-link--color-blue.nova-e-link--theme-bare',
        '.publication-meta-journal A',
        ],
      abstract: [
         '.publication-abstract .nova-e-text--spacing-auto',
         'DIV.nova-e-text.nova-e-text--size-m.nova-e-text--family-sans-serif.nova-e-text--spacing-auto.nova-e-text--color-inherit'
        ],
      authors: {
        selector: '.publication-author-list__item',
        subselectors: [
          { name: '.nova-v-person-list-item__title A' },
          { institution: 'LI.nova-v-person-list-item__meta-item:last-child SPAN' }
        ]
      }
    }
    scraper.scrape(res, selectors, '/publication/{id}', id, req.url)
  })

  /**
   * @swagger
   * /rg/author:
   *   get:
   *     description: "Returns ResearchGate author info."
   *     summary: "Get ResearchGate author info"
   *     tags: [ResearchGate]
   *     parameters:
   *       - name: name
   *         in: query
   *         description: "The Author name"
   *         required: true
   *         default: Jordan_Peterson2
   *         type: string
   *         consumes:
   *           - application/json
   *           - application/text
   *     produces:
   *       - application/json
   *     responses:
   *       200:
   *         headers:
   *           content-type: "text/plain"
   */
  server.get('/rg/author', function (req, res, next) {
    const id = req.query.name
    const selectors = {
      score: 'DIV[title="RG Score"]',
      title: 'SPAN.title',
      name: 'H1 SPAN',
      institution: '.info-header A:first-child',
      department: '.info-header A:nth-child(2)',
      researchItems: 'DIV.section-about>DIV:nth-child(2) .nova-o-grid.nova-o-grid--gutter-xxl.nova-o-grid--order-normal.nova-o-grid--horizontal-align-left.nova-o-grid--vertical-align-top DIV>DIV',
      reads: 'DIV.section-about>DIV:nth-child(2) .nova-o-grid.nova-o-grid--gutter-xxl.nova-o-grid--order-normal.nova-o-grid--horizontal-align-left.nova-o-grid--vertical-align-top DIV:nth-child(2)>DIV',
      citations: 'DIV.section-about>DIV:nth-child(2) .nova-o-grid.nova-o-grid--gutter-xxl.nova-o-grid--order-normal.nova-o-grid--horizontal-align-left.nova-o-grid--vertical-align-top DIV:nth-child(3)>DIV'
    }
    scraper.scrape(res, selectors, '/profile/{id}', id, req.url)
  })

  /**
   * @swagger
   * /rg/articles:
   *   get:
   *     description: "Returns article info for a given author."
   *     summary: "Get ResearchGate article list"
   *     tags: [ResearchGate]
   *     parameters:
   *       - name: name
   *         in: query
   *         description: "The Author name"
   *         required: true
   *         default: Jordan_Peterson2
   *         type: string
   *         consumes:
   *           - application/json
   *           - application/text
   *     produces:
   *       - application/json
   *     responses:
   *       200:
   *         headers:
   *           content-type: "text/plain"
   */
  server.get('/rg/articles', function (req, res, next) {
    const id = req.query.name
    const selectors = {
      articles: {
        selector: 'DIV[itemtype="http://schema.org/ScholarlyArticle"]',
        subselectors: [
          { title: 'A.nova-e-link' },
          { abstract: 'DIV[itemprop="description"]' },
          { type: 'A.nova-v-publication-item__type' },
          { date: 'LI.publication-item-meta-items__meta-data-item SPAN' },
          { reads: 'LI.nova-v-publication-item__metrics-item SPAN' },
          { authors: {
            selector: 'UL.nova-v-publication-item__person-list LI',
            subselectors: [
              { author: 'A[itemprop="author"]' }
            ]
            }
          }
        ]
      }
    }
    scraper.scrape(
      res,
      selectors,
      ['/profile/{id}'], //'/profile/{id}/2', '/profile/{id}/3'],
      id,
      req.url)
  })

  /**
   * @swagger
   * /rg/journal:
   *   get:
   *     description: "Returns info for a given journal."
   *     summary: "Get ResearchGate article list"
   *     tags: [ResearchGate]
   *     parameters:
   *       - name: name
   *         in: query
   *         description: "The Journal name"
   *         required: true
   *         default: 0138-9130_Scientometrics
   *         type: string
   *         consumes:
   *           - application/json
   *           - application/text
   *     produces:
   *       - application/json
   *     responses:
   *       200:
   *         headers:
   *           content-type: "text/plain"
   */
  server.get('/rg/journal', function (req, res, next) {
    const id = req.query.name
    const selectors = {
      publisher: 'DIV.journal-full-info__meta',
      description: '.journal-full-info__description-title + p',
      rgimpact: 'H2.nova-e-text--color-inherit b',
      halflife: '//x:th[contains(text(),\'Cited half-life\')]/parent::*/x:td',
      immediacy: '//x:th[contains(text(),\'Immediacy index\')]/parent::*/x:td',
      eigenfactor: '//x:th[contains(text(),\'Eigenfactor\')]/parent::*/x:td',
      influence: '//x:th[contains(text(),\'Article influence\')]/parent::*/x:td',
      website: '//x:th[contains(text(),\'Website\')]/parent::*/x:td',
      issn: '//x:th[contains(text(),\'ISSN\')]/parent::*/x:td',
      material: '//x:th[contains(text(),\'Material type\')]/parent::*/x:td',
      doctype: '//x:th[contains(text(),\'Document type\')]/parent::*/x:td',
    }
    scraper.scrape(
      res,
      selectors,
      ['/journal/{id}'], //'/profile/{id}/2', '/profile/{id}/3'],
      id,
      req.url)
  })

  /**
   * @swagger
   * /rg/citations:
   *   get:
   *     description: "Returns citation list for a given article."
   *     summary: "Get ResearchGate article list"
   *     tags: [ResearchGate]
   *     parameters:
   *       - name: title
   *         in: query
   *         description: "The article title."
   *         required: true
   *         default: 269286890_Openness_to_Experience_and_Intellect_Differentially_Predict_Creative_Achievement_in_the_Arts_and_Sciences_Openness_Intellect_and_Creativity
   *         type: string
   *         consumes:
   *           - application/json
   *           - application/text
   *     produces:
   *       - application/json
   *     responses:
   *       200:
   *         headers:
   *           content-type: "text/plain"
   */
  server.get('/rg/citations', function (req, res, next) {
    const id = req.query.title
    const selectors = {
      citations: {
        selector: '.pub-citations-list LI.pub-citations-list__item',
        subselectors: [
          {title: 'A.nova-e-link'},
          {cit: 'DIV.nova-v-citation-item__context-body--clamp-none'},
        ]
      }
    }
    scraper.scrape(res, selectors, '/publication/{id}', id, req.url)
  })

}
